Replace Cirros with Alpine

AlonaKaplan opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:

Cirros has a very slow startup in case there is no DHCP server responding to it.
To allow using the same tests for dual stack and IPv6 only clusters, tests using Cirros will now use Alpine.


**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
In this PR Cirros tests that run in IPv6 only cluster were switched to use the existing Alpine image.
The existing Alpine image needs some configuration for the network work. Cloud init is not installed, so the console has to be used for the configuration.

A follow up PR will use a new Alpine image that includes cloud-init so the manual console configuration will be removed,

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

AlonaKaplan committed:
tests: Activate VM interfaces and run dhcp server

Cirros has a very slow startup in case there is no dhcp server responding
to it. To allow using the same tests for dual stack and ipv6 only clusters,
tests using Cirros will now use Alpine.

Alpine VM starts with eth0 and lo interfaces down.
This commit is adding activation to those interfaces and running the dhcp
server.

Signed-off-by: Alona Kaplan <alkaplan@redhat.com>

AlonaKaplan committed:
tests: Replace Cirros with Alpine

Cirros has a very slow startup in case there is no DHCP server responding
to it. To allow using the same tests for dual stack and IPv6 only clusters,
tests using Cirros will now use Alpine.

Signed-off-by: Alona Kaplan <alkaplan@redhat.com>

AlonaKaplan committed:
tests: Decrease timeout to login Cirros

The timeout was increased as a workarround to ipv6 only cluster
(the startup of a Cirros VM in case the is no dhcp server is very long).

Now the Cirros VMs were substituted with alpine the workarround can be
removed.

Note: the commit the timeout inclease logic from cirros login as well.
This logic was added there by mistake from the first place.


Signed-off-by: Alona Kaplan <alkaplan@redhat.com>

AlonaKaplan commented:
/test pull-kubevirt-e2e-k8s-1.22-ipv6-sig-network

AlonaKaplan commented:
/test pull-kubevirt-e2e-k8s-1.22-sig-network

rmohr commented:
> Cirros has a very slow startup in case there is no DHCP server responding to it.

One maybe dump question: I would assume that, while potentially routing information and autoconfigured ipv6 works to a degree, I would expect that we still have DHCPv6 to provide the full set of information (like host name, ...). Otherwise ipv6 seems to be a little bit of an island solution to me.

@AlonaKaplan any thoughts on that?

AlonaKaplan commented:
> > Cirros has a very slow startup in case there is no DHCP server responding to it.
> 
> One maybe dump question: I would assume that, while potentially routing information and autoconfigured ipv6 works to a degree, I would expect that we still have DHCPv6 to provide the full set of information (like host name, ...). Otherwise ipv6 seems to be a little bit of an island solution to me.
> 
> @AlonaKaplan any thoughts on that?

Hi @rmohr ,
In this specific case even if virt-launcher provided DHCP6 + RA it wouldn't be enough since Cirros doesn't have a properly working DHCP6 client (last time I tested `odhcp6c` it wasn't working).
On IPv6 only cluster the start-up is slowed down by `/sbin/cirros-dhcpc` discover waiting for a response. `/sbin/cirros-dhcpc` is IPv4 only client, so even even we had DHCP6 server + RA I believe it wouldn't help.

>  I would expect that we still have DHCPv6 to provide the full set of information (like host name, ...). Otherwise ipv6 seems to be a little bit of an island solution to me.

We have [DHCP server]( https://github.com/kubevirt/kubevirt/blob/cd108b3d8edd3ecd57fdad62f98a8d6c0788db2a/pkg/network/dhcp/serverv6/serverv6.go).
In IPv6, first router solicitation is sent by the client machine, only once the client receives router advertisement (and the advertisement specifies the DHCP6 client is needed) the DHCP6 client is sending is a request.

We don't have RA. We worked on it, but the [PR](https://github.com/kubevirt/kubevirt/pull/4581) was eventually closed due to low priority.
For a dual stack cluster the IPv4 DHCP server provided all the extra information (host name, dns server, search domains etc.).
Now we support single stack IPv6 I agree that it may worth reviving the RA PR. It is in our TODO list to consider it:)
@phoracek

rmohr commented:
> Hi @rmohr ,
> In this specific case even if virt-launcher provided DHCP6 + RA it wouldn't be enough since Cirros doesn't have a properly working DHCP6 client (last time I tested `odhcp6c` it wasn't working).
> On IPv6 only cluster the start-up is slowed down by `/sbin/cirros-dhcpc` discover waiting for a response. `/sbin/cirros-dhcpc` is IPv4 only client, so even even we had DHCP6 server + RA I believe it wouldn't help.

Yes I have seen bug reports regarding to `/sbin/cirros-dhcpc`. I would of course prefer to get this working to stay with very fast boots and the lowest possible memory footprint. I would also not mind having a custom cirros build. It would be a setback for the test suite. :/

That is basically my only concern, otherwise I don't care too much which image is used.

Thanks for clarifying where we stand with RA and dhcpv6.

rmohr commented:
> From my experiments it seems to be as fast as Cirros. I didn't check about the memory footprint.

Just checked. I though we take 256Mi of memory, but it is 128Mi too. If it boots within the same amount of time it may be good to switch.

AlonaKaplan commented:
/test pull-kubevirt-e2e-k8s-1.22-ipv6-sig-network

AlonaKaplan commented:
/test pull-kubevirt-e2e-k8s-1.22-sig-network

AlonaKaplan commented:
/test pull-kubevirt-e2e-k8s-1.22-ipv6-sig-network

AlonaKaplan commented:
/test pull-kubevirt-e2e-k8s-1.22-ipv6-sig-network

AlonaKaplan commented:
/test pull-kubevirt-e2e-k8s-1.22-sig-network

AlonaKaplan commented:
/test pull-kubevirt-e2e-k8s-1.22-ipv6-sig-network

AlonaKaplan commented:
/test pull-kubevirt-e2e-k8s-1.22-sig-network
/test pull-kubevirt-e2e-k8s-1.22-ipv6-sig-network

AlonaKaplan commented:
/test pull-kubevirt-e2e-k8s-1.22-sig-network
/test pull-kubevirt-e2e-k8s-1.22-ipv6-sig-network

AlonaKaplan commented:
/test pull-kubevirt-e2e-k8s-1.22-sig-network
/test pull-kubevirt-e2e-k8s-1.22-ipv6-sig-network

AlonaKaplan committed:
tests: Used wget instead of curl on Alpine VMs

The alpine image we currently use in our e2e tests doesn't include `curl`.
`wget` is used instead.

Signed-off-by: Alona Kaplan <alkaplan@redhat.com>

AlonaKaplan committed:
tests: changes probes tests to use a masquerade binding

The tests are testing both IPv4 and IPv6, however, IPv6 is not supported by
bridge binding so the communcation was with the virt-launcher pod and
not the VM.

Signed-off-by: Alona Kaplan <alkaplan@redhat.com>

AlonaKaplan commented:
/test pull-kubevirt-e2e-k8s-1.22-sig-network
/test pull-kubevirt-e2e-k8s-1.22-ipv6-sig-network

AlonaKaplan committed:
tests: Rename libnet.WithIPv6 to libnet.WithAlpineConfig

Signed-off-by: Alona Kaplan <alkaplan@redhat.com>

AlonaKaplan commented:
/test pull-kubevirt-e2e-k8s-1.22-sig-network
/test pull-kubevirt-e2e-k8s-1.22-ipv6-sig-network

AlonaKaplan commented:
/test pull-kubevirt-e2e-k8s-1.22-sig-network
/test pull-kubevirt-e2e-k8s-1.22-ipv6-sig-network

qinqon reviewed (changes requested):
Just one little question

AlonaKaplan commented on tests/libnet/ipv6.go:
```diff
@@ -12,42 +12,72 @@ import (
-	hasEth0Iface := func() bool {
-		err := console.RunCommand(vmi, "ip a | grep -q eth0", 30*time.Second)
+	alreadyConfigured := func() bool {
+		err := console.RunCommand(vmi, " ip a show lo | grep UP", 30*time.Second)
```
No, it is `lo` intentionally. Since we may have a VM with no eth0 at all (in case `AutoattachPodInterface` is false and the spec has no interfaces).
Since the `if` checking `alreadyConfigured` also checking if `AutoattachPodInterface` is false, we may use `eth0`. But I prefer to keep `lo` to be on the safe side.

qinqon commented:
/lgtm

AlonaKaplan commented:
/retest

AlonaKaplan committed:
tests: Remove `WithAlpineConfig` for non Alpine VMs.

Signed-off-by: Alona Kaplan <alkaplan@redhat.com>

AlonaKaplan commented:
/test pull-kubevirt-e2e-k8s-1.22-sig-network
/test pull-kubevirt-e2e-k8s-1.22-ipv6-sig-network

AlonaKaplan committed:
tests: skip IPv4 only tests on IPv6 only cluster if IPv4 is not supported

The skipped tests are currently checking only the IPv4 connectivity.
An issue was opened to adapt them for IPv6 as well.

Issue- https://github.com/kubevirt/kubevirt/issues/7506

Signed-off-by: Alona Kaplan <alkaplan@redhat.com>

AlonaKaplan commented:
/test pull-kubevirt-e2e-k8s-1.22-sig-network
/test pull-kubevirt-e2e-k8s-1.22-ipv6-sig-network

AlonaKaplan commented:
/test pull-kubevirt-e2e-k8s-1.22-sig-network
/test pull-kubevirt-e2e-k8s-1.22-ipv6-sig-network

AlonaKaplan commented:
/retest-required

AlonaKaplan commented:
/retest

AlonaKaplan commented:
/retest

AlonaKaplan commented:
/retest

AlonaKaplan commented:
/retest

AlonaKaplan commented:
/retest

phoracek commented on tests/network/networkpolicy.go:
```diff
@@ -388,10 +388,10 @@ func checkHTTPPingAndStopOnFailure(fromVmi, toVmi *v1.VirtualMachineInstance, po
 
 func checkHTTPPing(vmi *v1.VirtualMachineInstance, ip string, port int) error {
-	const curlCheckCmd = "curl --head %s --connect-timeout 5\n"
+	const wgetCheckCmd = "wget -S --spider %s -T 5\n"
```
wget adds way more noise than curl, are we ok with this change?

```
[phoracek@t490s ~]$ wget -S --spider google.com -T 5 2>&1 | wc -l
37

[phoracek@t490s ~]$ curl --head google.com --connect-timeout 5 2>&1 | wc -l
14
```

AlonaKaplan replied:
alpine doesn't have `curl` so that's the best option we have. Why should we care about this "noise"?

AlonaKaplan replied:
well, unfortunately I don't see any other option with alpine.

phoracek replied:
All good then, feel free to cancel the hold any time

phoracek reviewed (commented):
Minor nit. Feel free to cancel the hold once ready.

There were only rebases since Quique's LGTM.

/lgtm
/approve

AlonaKaplan commented:
/retest

xuzhenglun commented:
> Now we support single stack IPv6 I agree that it may worth reviving the RA PR. It is in our TODO list to consider it:)

sorry for replying a closed issue, I'm wounding if any update for reviving the RA PR.  I'm not asking a feature request, just want to know about everyone's recent progress and opinions on this matter. :)

@AlonaKaplan
